File: Location.cpp
class Location {
public:
  string filename;
  int line;
  int position;
  Location() :
    filename("<internal>"), line(0), position(0) {}
  Location(string filename) :
    filename(filename), line(0), position(0) {}
  Location(string filename, int line, int position) :
    filename(filename), line(line), position(position) {}
  void advanceNextLine() { line++; position=0; }
  void advanceNextCharacter() { position++; }
  string toString() {
    return filename + ":" + intToString(line+1) + ":" + intToString(position+1);
  }
};

File: src/Location.java
public class Location {
    String filename;
    int line;
    int position;
    Location() {
        this.filename = "<internal>";
    }
    Location(String filename) {
        this.filename = filename;
    }
    Location(String filename, int line, int position) {
        this.filename = filename;
        this.line = line;
        this.position = position;
    }
    Location(Location location) {
        this.filename = location.filename;
        this.line = location.line;
        this.position = location.position;
    }
    public String toString() {
        return filename + ":" + Integer.toString(line+1) + ":" + Integer.toString(position+1);
    }
    void advanceNextLine() { line++; position = 0; }
    void advanceNextCharacter() { position++; }
}
